### PR TITLE
[X-PR] p4est edge operations and consistent corner operation

### PR DESCRIPTION
**Pull request by koehlerson**
- [x] intraoctree
    - [x] `edge_neighbor`
    - [x] tests
    - [x] docs 
- [x] interoctree
    - [x] remote version that implements exactly the logic from the paper 
    - [x] version with flipped logic that seems more natural (analogous to `transform_face`)
    - [x] tests
    - [x] docs
- [x] high level functionality
    - [x] balancing
    - [x] creategrid
    - [x] hanging nodes

### CODE REVIEW

File: test/test_p4est.jl
[the remark this answers is not in the document]
Comment (koehlerson) [reply test/test_p4est.jl:1]: yeah definitely

File: src/Adaptivity/AdaptiveCells.jl [hunk 1]
```diff
@@ -46,7 +46,7 @@ end
 #OctantBWG(dim::Int,l::Int,m::Int,b::Int=_maxlevel[dim-1]) = OctantBWG(dim,l,m,b)
 #OctantBWG(dim::Int,l::Int,m::Int,b::Int32) = OctantBWG(dim,l,m,b)
 #OctantBWG(dim::Int,l::Int32,m::Int,b::Int32) = OctantBWG(dim,l,Int32(m),b)
-function OctantBWG(level::Int,coords::NTuple)
+function OctantBWG(level::T,coords::NTuple) where T <: Integer
```
Comment (termi-official) [inline src/Adaptivity/AdaptiveCells.jl:49]:
```suggestion
function OctantBWG(level::Integer,coords::NTuple)
```
Is there a reason why you grab T explicitly?

**Review by termi-official — COMMENTED** [review]
I think we should start adding some docs to the code to understand what we have created here :D

File: src/Adaptivity/AdaptiveCells.jl (hunk 1 above)
Comment (koehlerson) [reply src/Adaptivity/AdaptiveCells.jl:49]: I thought that we could add later a check if `level` and `eltype(coords)` matches and use the wider/smaller type

File: src/Adaptivity/AdaptiveCells.jl [hunk 2]
```diff
@@ -614,31 +616,71 @@ function creategrid(forest::ForestBWG{dim,C,T}) where {dim,C,T}
...
         end
         if dim > 2
-            #TODO add egde duplication check
-            @error "Edge deduplication not implemented yet."
+            _edges = edges(root(dim),tree.b)
+            # Face neighbors
+            @debug println("Updating edge neighbors for octree $k")
+            for (e,ec) in enumerate(_edges) # f in p4est notation
+                e_axis_index, e_axis_sign = divrem(e-1,4) #first axis 0 (x), 1 (y), 2(z), second positive or negative direction
+                edge_neighbor = forest.topology.edge_edge_neighbor[k,edge_perm[e]]
```
[the remark this answers is not in the document]
Comment (koehlerson) [reply src/Adaptivity/AdaptiveCells.jl:670]: TODO for later PR

File: src/Adaptivity/AdaptiveCells.jl [hunk 3]
```diff
@@ -741,11 +801,59 @@ function hangingnodes(forest::ForestBWG{dim}, nodeids, nodeowners) where dim
...
+                                end
+                                if contains_edge(re, pedge)
+                                    k′ = edge_neighbor[1][1]
+                                    ri′ = edge_perm_inv[edge_neighbor[1][2]]
+                                    interoctree_neighbor = transform_edge(forest, k′, ri′, neighbor_candidate, true)
+                                    interoctree_neighbor_candidate_idx = findfirst(x->x==interoctree_neighbor,forest.cells[k′].leaves)
+                                    if interoctree_neighbor_candidate_idx !== nothing
+                                        neighbor_candidate_edges = edges(neighbor_candidate,forest.cells[k′].b)
+                                        transformed_neighbor_edges = edges(interoctree_neighbor,forest.cells[k′].b)
+                                        ne = findfirst(x->x==pedge,neighbor_candidate_edges) #TODO needs rotation stuff I think
```
[the remark this answers is not in the document]
Comment (koehlerson) [reply src/Adaptivity/AdaptiveCells.jl]: as discussed relaxed the requirement from `x->x==pedge` to `x->iscenter(c,x)` which should be rotational invariant. Test should already be there in the test suite

**Comment by termi-official** [comment]
Seems like rebalance over edges is still broken. MWE:

Screenshot from 2024-04-24 17-12-38

```julia
using Ferrite, FerriteGmsh, SparseArrays
grid = generate_grid(Hexahedron, (3,3,2));
grid  = ForestBWG(grid,10)

function adapt_mesh_random_sequence(initial_grid, nctor = 5)
    i = 1
    grid = deepcopy(initial_grid)
    pvd = paraview_collection("amr-grid.pvd");
    while i<=10
        @show i
        transfered_grid = Ferrite.creategrid(grid)
        vtk_grid("amr-grid_$i", transfered_grid) do vtk
            pvd[i] = vtk
        end

        dh = DofHandler(transfered_grid)
        add!(dh, :u, Lagrange{RefHexahedron, 1}())
        close!(dh);

        ch = ConstraintHandler(dh)
        add!(ch, ConformityConstraint(:u))
        close!(ch);

        cells_to_refine = rand(1:length(transfered_grid.cells), nctor)
        Ferrite.refine!(grid, cells_to_refine)
        Ferrite.balanceforest!(grid)

        i += 1
    end
    vtk_save(pvd);
end

adapt_mesh_random_sequence(grid)
```

**Comment by koehlerson** [comment]
grafik
looks correctly balanced to me, but throws nested affine constraints

**Comment by termi-official** [comment]
Dumping this one for later

```julia
using Ferrite, FerriteGmsh, SparseArrays
grid = generate_grid(Hexahedron, (3,3,2));
function random_deformation_field(x::Vec{sdim}) where sdim
    if any(x .≈ -1.0) || any(x .≈ 1.0)
        return x
    else
        Vec{sdim}(x .+ (rand(sdim).-0.5)*0.15)
    end
end
transform_coordinates!(grid, random_deformation_field)
grid  = ForestBWG(grid,10)

analytical_solution(x) = atan(2*(norm(x)-0.5)/0.02)
analytical_rhs(x) = -laplace(analytical_solution,x)

function assemble_cell!(ke, fe, cellvalues, ue, coords)
    fill!(ke, 0.0)
    fill!(fe, 0.0)

    n_basefuncs = getnbasefunctions(cellvalues)
    for q_point in 1:getnquadpoints(cellvalues)
        x = spatial_coordinate(cellvalues, q_point, coords)
        dΩ = getdetJdV(cellvalues, q_point)
        for i in 1:n_basefuncs
            Nᵢ = shape_value(cellvalues, q_point, i)
            ∇Nᵢ = shape_gradient(cellvalues, q_point, i)# shape_symmetric_gradient(cellvalues, q_point, i)
            fe[i] += analytical_rhs(x) * Nᵢ * dΩ # add internal force to residual
            for j in 1:n_basefuncs
                ∇Nⱼ = shape_gradient(cellvalues, q_point, j)
                ke[i, j] += ∇Nⱼ ⋅ ∇Nᵢ * dΩ
            end
        end
    end
end

function assemble_global!(K, f, a, dh, cellvalues)
    ## Allocate the element stiffness matrix and element force vector
    n_basefuncs = getnbasefunctions(cellvalues)
    ke = zeros(n_basefuncs, n_basefuncs)
    fe = zeros(n_basefuncs)
    ## Create an assembler
    assembler = start_assemble(K, f)
    ## Loop over all cells
    for cell in CellIterator(dh)
        reinit!(cellvalues, cell) # update spatial derivatives based on element coordinates
        @views ue = a[celldofs(cell)]
        ## Compute element contribution
        coords = getcoordinates(cell)
        assemble_cell!(ke, fe, cellvalues, ue, coords)
        ## Assemble ke and fe into K and f
        assemble!(assembler, celldofs(cell), ke, fe)
    end
    return K, f
end

function solve(grid)
    dim = 2
    order = 1 # linear interpolation
    ip = Lagrange{RefHexahedron, order}() # vector valued interpolation
    qr = QuadratureRule{RefHexahedron}(2) # 1 quadrature point
    cellvalues = CellValues(qr, ip);

    dh = DofHandler(grid)
    add!(dh, :u, ip)
    close!(dh);

    ch = ConstraintHandler(dh)
    add!(ch, ConformityConstraint(:u))
    add!(ch, Dirichlet(:u, getfaceset(grid, "top"), (x, t) -> 0.0))
    add!(ch, Dirichlet(:u, getfaceset(grid, "right"), (x, t) -> 0.0))
    add!(ch, Dirichlet(:u, getfaceset(grid, "left"), (x, t) -> 0.0))
    add!(ch, Dirichlet(:u, getfaceset(grid, "bottom"), (x, t) -> 0.0))
    close!(ch);

    K = create_sparsity_pattern(dh,ch)
    f = zeros(ndofs(dh))
    a = zeros(ndofs(dh))
    assemble_global!(K, f, a, dh, cellvalues);
    apply!(K, f, ch)
    u = K \ f;
    apply!(u,ch)
    return u,dh,ch,cellvalues
end

function compute_fluxes(u,dh::DofHandler{sdim}) where sdim
    ip = Lagrange{RefHexahedron, 1}()
    # Normal quadrature points
    qr = QuadratureRule{RefHexahedron}(2)
    cellvalues = CellValues(qr, ip);
    # Superconvergent point
    qr_sc = QuadratureRule{RefHexahedron}(1)
    cellvalues_sc = CellValues(qr_sc, ip);
    #Buffers
    σ_gp = Vector{Vector{Vec{sdim,Float64}}}()
    σ_gp_loc = Vector{Vec{sdim,Float64}}()
    σ_gp_sc = Vector{Vector{Vec{sdim,Float64}}}()
    σ_gp_sc_loc = Vector{Vec{sdim,Float64}}()
    for (cellid,cell) in enumerate(CellIterator(dh))
        @views ue = u[celldofs(cell)]

        reinit!(cellvalues, cell)
        for q_point in 1:getnquadpoints(cellvalues)
            gradu = function_gradient(cellvalues, q_point, ue)
            push!(σ_gp_loc, gradu)
        end
        push!(σ_gp,copy(σ_gp_loc))
        empty!(σ_gp_loc)

        reinit!(cellvalues_sc, cell)
        for q_point in 1:getnquadpoints(cellvalues_sc)
            gradu = function_gradient(cellvalues_sc, q_point, ue)
            push!(σ_gp_sc_loc, gradu)
        end
        push!(σ_gp_sc,copy(σ_gp_sc_loc))
        empty!(σ_gp_sc_loc)
    end
    return σ_gp, σ_gp_sc
end

function solve_adaptive(initial_grid::Ferrite.AbstractGrid{sdim}) where sdim
    ip = Lagrange{RefHexahedron, 1}()^sdim
    qr_sc = QuadratureRule{RefHexahedron}(1)
    cellvalues_flux = CellValues(qr_sc, ip);
    finished = false
    i = 1
    grid = deepcopy(initial_grid)
    pvd = paraview_collection("heat_amr.pvd");
    while !finished && i<=10
        @show i
        transfered_grid = Ferrite.creategrid(grid)
        vtk_grid("heat_amr-grid_$i", transfered_grid) do vtk
        end
        u,dh,ch,cv = solve(transfered_grid)
        σ_gp, σ_gp_sc = compute_fluxes(u,dh)
        projector = L2Projector(Lagrange{RefHexahedron, 1}(), transfered_grid)
        σ_dof = project(projector, σ_gp, QuadratureRule{RefHexahedron}(2))
        cells_to_refine = Int[]
        error_arr = Float64[]
        for (cellid,cell) in enumerate(CellIterator(projector.dh))
            reinit!(cellvalues_flux, cell)
            @views σe = σ_dof[celldofs(cell)]
            error = 0.0
            for q_point in 1:getnquadpoints(cellvalues_flux)
                σ_dof_at_sc = function_value(cellvalues_flux, q_point, σe)
                error += norm((σ_gp_sc[cellid][q_point] - σ_dof_at_sc ))
                error *= getdetJdV(cellvalues_flux,q_point)
            end
            if error > 0.001
                push!(cells_to_refine,cellid)
            end
            push!(error_arr,error)
        end

        vtk_grid("heat_amr-iteration_$i", dh) do vtk
            vtk_point_data(vtk, dh, u)
            vtk_point_data(vtk, projector, σ_dof, "flux")
            vtk_cell_data(vtk, getindex.(collect(Iterators.flatten(σ_gp_sc)),1), "flux sc x")
            vtk_cell_data(vtk, getindex.(collect(Iterators.flatten(σ_gp_sc)),2), "flux sc y")
            vtk_cell_data(vtk, error_arr, "error")
            pvd[i] = vtk
        end

        Ferrite.refine!(grid, cells_to_refine)
        Ferrite.balanceforest!(grid)

        i += 1
        if isempty(cells_to_refine)
            finished = true
        end
    end
    vtk_save(pvd);
end

solve_adaptive(grid)

```

**Pull request merged**